include/revng/Pipeline/Containers.h

```diff
@@ -0,0 +1,227 @@
+
+  const ContainerBase &get() const { return **Container; }
+
+  ContainerBase &getOrCreate() {
+    if (not hasValue())
+      Container = Factory->createEmpty();
+
+    return get();
+  }
+
+  bool hasValue() const { return Container.has_value(); }
+
+  void mergeBack(MaybeContainer &&Other) {
+    if (not Other.hasValue())
+      return;
+
+    if (not hasValue()) {
+      *this = std::move(Other);
+      return;
+    }
+
+    get().mergeBack(std::move(Other.get()));
+  }
+
+  std::unique_ptr<ContainerBase> extract() {
+    if (not hasValue())
+      return nullptr;
+
+    auto ptr = std::move(*Container);
+    Container = std::nullopt;
+    return ptr;
+  }
+
+  MaybeContainer cloneFiltered(const TargetsList &Targets) const {
+    if (not hasValue())
+      return MaybeContainer(*Factory);
+    return MaybeContainer(*Factory, get().cloneFiltered(Targets));
+  }
+
+  llvm::Error storeToDisk(llvm::StringRef Directory) const {
+    if (not hasValue())
+      return llvm::Error::success();
+
+    return get().storeToDisk(Directory);
+  }
+
+  llvm::Error loadFromDisk(llvm::StringRef File) {
+    if (not llvm::sys::fs::exists(File)) {
+      Container = std::nullopt;
+      return llvm::Error::success();
+    }
+
+    return getOrCreate().loadFromDisk(File);
+  }
+
+  TargetsList enumerate() const {
+    if (not hasValue())
+      return {};
+
+    return get().enumerate();
+  }
+};
+
+/// a map from containerName to a MaybeContainer, which is initially empty.
+class Containers {
+public:
+  using Map = llvm::StringMap<MaybeContainer>;
+  using const_iterator = Map::const_iterator;
+
+private:
+  Map Content;
+
+public:
+  Containers(Map Content) : Content(std::move(Content)) {}
+  Containers() = default;
+  Containers(Containers &&) = default;
+  Containers &operator=(Containers &&) = default;
+  ~Containers() = default;
+  Containers(const Containers &) = delete;
+  Containers &operator=(const Containers &) = delete;
+
+  Containers cloneFiltered(const ContainersStatus &Targets);
+
+  void mergeContainers(Containers &&Other) {
+    for (auto &Entry : Other.Content) {
+      revng_assert(Content.count(Entry.first()) != 0);
+      MaybeContainer &LContainer = Content.find(Entry.first())->second;
+      MaybeContainer &RContainer = Entry.second;
+      LContainer.mergeBack(std::move(RContainer));
+    }
+  }
+
+  template<typename ContainerDerivedType>
+  ContainerDerivedType &getOrCreate(llvm::StringRef Name) {
+    revng_assert(containsOrCanCreate(Name));
+    auto &Base = Content.find(Name)->second.getOrCreate();
+    revng_assert(llvm::isa<ContainerDerivedType>(Base));
+    return llvm::cast<ContainerDerivedType>(Base);
+  }
+
+  template<typename ContainerDerivedType>
+  const ContainerDerivedType &get(llvm::StringRef Name) const {
+    revng_assert(contains(Name));
+    const auto &Base = Content.find(Name)->second.get();
+    revng_assert(llvm::isa<ContainerDerivedType>(Base));
+    return llvm::cast<ContainerDerivedType>(Base);
+  }
```

Comment:
Uhm, it looks like to me that it's the user responsibility to either `llvm::cast` or maybe `llvm::dyn_cast`?

Reply:
mh i wouldn't say that because there is no particular reason to ask back a container in then you are not able to do anything with it. You would end up casting it up 99% of the times anyway

Reply:
OK, then why expose also the other `get`? Maybe make them private.
If you want your API to be opinionated about its usage, go for it, but let's not leave a super wide API where you have multiple ways to do the same thing.

Reply:
Also, the upcasting versions should use the non-upcasting ones, otherwise we're needlessly duplicating the implementation.